File: Software/src/battery/SAMSUNG-SDI-LV-BATTERY.cpp
#include "SAMSUNG-SDI-LV-BATTERY.h"
#include "../communication/can/comm_can.h"
#include "../datalayer/datalayer.h"
#include "../devboard/utils/events.h"

/*
- Baud rate: 500kbps
- Format: CAN2.0A 11 bit identifier
- Data Length: 8byte
- CAN data is transmitted with encoding in little endian – low byte first – unless stated otherwise.
- Broadcasting period: 500ms

TODO: Implement the error bit handling for easier visualization if the battery stops operating
- alarms_frame0
- alarms_frame1
- protection_frame2
- protection_frame3

*/

void SamsungSdiLVBattery::update_values() {

  datalayer.battery.status.real_soc = system_SOC * 100;

  datalayer.battery.status.remaining_capacity_Wh = static_cast<uint32_t>(
      (static_cast<double>(datalayer.battery.status.real_soc) / 10000) * datalayer.battery.info.total_capacity_Wh);

  datalayer.battery.status.soh_pptt = system_SOH * 100;

  datalayer.battery.status.voltage_dV = system_voltage / 10;

  datalayer.battery.status.current_dA = system_current * 10;

  datalayer.battery.status.max_charge_power_W = system_voltage * charge_current_limit;

  datalayer.battery.status.max_discharge_power_W = system_voltage * discharge_current_limit;

  datalayer.battery.status.temperature_min_dC = (int16_t)(minimum_cell_temperature * 10);

  datalayer.battery.status.temperature_max_dC = (int16_t)(maximum_cell_temperature * 10);

  datalayer.battery.status.cell_max_voltage_mV = maximum_cell_voltage;

  datalayer.battery.status.cell_min_voltage_mV = minimum_cell_voltage;

  datalayer.battery.info.max_design_voltage_dV = battery_charge_voltage;

  datalayer.battery.info.min_design_voltage_dV = battery_discharge_voltage;
}

void SamsungSdiLVBattery::handle_incoming_can_frame(CAN_frame rx_frame) {
  switch (rx_frame.ID) {
    case 0x500:  //Voltage, current, SOC, SOH
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      system_voltage = ((rx_frame.data.u8[0] << 8) | rx_frame.data.u8[1]);
      system_current = ((rx_frame.data.u8[2] << 8) | rx_frame.data.u8[3]);
      system_SOC = rx_frame.data.u8[4];
      system_SOH = rx_frame.data.u8[5];
      break;
    case 0x501:
      alarms_frame0 = rx_frame.data.u8[0];
      alarms_frame1 = rx_frame.data.u8[1];
      protection_frame2 = rx_frame.data.u8[2];
      protection_frame3 = rx_frame.data.u8[3];
      break;
    case 0x502:
      battery_charge_voltage = ((rx_frame.data.u8[0] << 8) | rx_frame.data.u8[1]);
      charge_current_limit = ((rx_frame.data.u8[2] << 8) | rx_frame.data.u8[3]);
      discharge_current_limit = ((rx_frame.data.u8[4] << 8) | rx_frame.data.u8[5]);
      battery_discharge_voltage = ((rx_frame.data.u8[6] << 8) | rx_frame.data.u8[7]);
      break;
    case 0x503:
      maximum_cell_voltage = ((rx_frame.data.u8[2] << 8) | rx_frame.data.u8[3]);
      minimum_cell_voltage = ((rx_frame.data.u8[4] << 8) | rx_frame.data.u8[5]);
      break;
    case 0x504:
      maximum_cell_temperature = rx_frame.data.u8[5];
      minimum_cell_temperature = rx_frame.data.u8[6];
      break;
    case 0x505:
      system_permanent_failure_status_dry_contact = rx_frame.data.u8[2];
      system_permanent_failure_status_fuse_open = rx_frame.data.u8[3];
      break;
    default:
      break;
  }
}

void SamsungSdiLVBattery::transmit_can(unsigned long currentMillis) {
  //No periodic sending required
}

void SamsungSdiLVBattery::setup(void) {  // Performs one time setup at startup
  strncpy(datalayer.system.info.battery_protocol, Name, 63);
  datalayer.system.info.battery_protocol[63] = '\0';
  datalayer.system.status.battery_allows_contactor_closing = true;
  datalayer.battery.info.max_design_voltage_dV = MAX_PACK_VOLTAGE_DV;
  datalayer.battery.info.min_design_voltage_dV = MIN_PACK_VOLTAGE_DV;
  datalayer.battery.info.max_cell_voltage_mV = MAX_CELL_VOLTAGE_MV;
  datalayer.battery.info.min_cell_voltage_mV = MIN_CELL_VOLTAGE_MV;
  datalayer.battery.info.total_capacity_Wh = 5000;
}

File: frontend/src/ext/datalayer/SamsungSdiLVBattery.ts
// SamsungSdiLVBattery: 52 bytes; base classes: CanBattery@0
export const SAMSUNG_SDI_LVBATTERY_FIELDS: ([string, string] | [string, string, number])[] = [
  ['___vptr$Battery', '__vtbl_ptr_type*'],
  ['__defaultRenderer', 'BatteryDefaultRenderer'],
  ['', ' ', 4],
  ['___vptr$Transmitter', '__vtbl_ptr_type*'],
  ['___vptr$CanReceiver', '__vtbl_ptr_type*'],
  ['_can_interface', 'CAN_Interface'],
  ['_initial_speed', 'CAN_Speed'],
  ['system_voltage', 'u16'],
  ['system_current', 'i16'],
  ['system_SOC', 'u8'],
  ['system_SOH', 'u8'],
  ['battery_charge_voltage', 'u16'],
  ['charge_current_limit', 'u16'],
  ['discharge_current_limit', 'u16'],
  ['battery_discharge_voltage', 'u16'],
  ['alarms_frame0', 'u8'],
  ['alarms_frame1', 'u8'],
  ['protection_frame2', 'u8'],
  ['protection_frame3', 'u8'],
  ['maximum_cell_voltage', 'u16'],
  ['minimum_cell_voltage', 'u16'],
  ['maximum_cell_temperature', 'i8'],
  ['minimum_cell_temperature', 'i8'],
  ['system_permanent_failure_status_dry_contact', 'u8'],
  ['system_permanent_failure_status_fuse_open', 'u8'],
];

export const ___vptr$Battery: [number, string] = [0, '__vtbl_ptr_type*'];
export const __defaultRenderer: [number, string] = [4, 'BatteryDefaultRenderer'];
export const ___vptr$Transmitter: [number, string] = [8, '__vtbl_ptr_type*'];
export const ___vptr$CanReceiver: [number, string] = [12, '__vtbl_ptr_type*'];
export const _can_interface: [number, string] = [16, 'CAN_Interface'];
export const _initial_speed: [number, string] = [20, 'CAN_Speed'];
export const system_voltage: [number, string] = [24, 'u16'];
export const system_current: [number, string] = [26, 'i16'];
export const system_SOC: [number, string] = [28, 'u8'];
export const system_SOH: [number, string] = [29, 'u8'];
export const battery_charge_voltage: [number, string] = [30, 'u16'];
export const charge_current_limit: [number, string] = [32, 'u16'];
export const discharge_current_limit: [number, string] = [34, 'u16'];
export const battery_discharge_voltage: [number, string] = [36, 'u16'];
export const alarms_frame0: [number, string] = [38, 'u8'];
export const alarms_frame1: [number, string] = [39, 'u8'];
export const protection_frame2: [number, string] = [40, 'u8'];
export const protection_frame3: [number, string] = [41, 'u8'];
export const maximum_cell_voltage: [number, string] = [42, 'u16'];
export const minimum_cell_voltage: [number, string] = [44, 'u16'];
export const maximum_cell_temperature: [number, string] = [46, 'i8'];
export const minimum_cell_temperature: [number, string] = [47, 'i8'];
export const system_permanent_failure_status_dry_contact: [number, string] = [48, 'u8'];
export const system_permanent_failure_status_fuse_open: [number, string] = [49, 'u8'];
